### PR TITLE
fix(ingest): a receiver typed with template arguments recorded no type, or the wrong class

### PR DESCRIPTION
**Pull request by joyful-ii-V-I**
## Sequencing — lands AFTER #256 and #268

Branched off `main` (13a19162, merged with fe28fd49). The fix is file-disjoint from both PRs in `src/`, but its biggest effect composes with #268's class identity (the inherited-member walk and `specializationDefs`), which the newly bound template receivers feed. `kParserVer` is declared **103** — past the 100–102 that the lanes queued ahead (#244, #243, #256) declare; the landing train assigns the real number and re-derives `test/qschemetrip.hash`. `test/narrowcheck.sh` appends arms 39–43 at the same place #268 appends 26–38, so the train merge should union the two blocks. `provOf`/`expectProv` carry #268's optional-map change byte for byte, so those lines merge clean.

## Defect

Rule 2 reads a receiver's type off its declaration (`src/ingest_binds.h` `writtenTypeOf`). It recorded a type only for a `type_identifier` or a qualified name:

1. **An unqualified template-id recorded nothing.** `SmallVectorImpl<FunctionDecl *> &v; v.push_back( FD )` never reached Rule 2 or class identity. Its qualified twin `llvm::SmallVectorImpl<…>&` did. This is how code inside `namespace llvm` / `ROCKSDB_NAMESPACE` writes almost every such type.
2. **The qualified path cut the spelling at its FIRST `<`** (`finalSegment`), so `Outer<int>::Inner& in` recorded `Outer`:
   - where `Outer` defines the method, a **precise edge to the wrong class** (fixture arm 41; no instance on either corpus);
   - otherwise a fall to the name ladder (86 `X<…>::Y` receivers change target: 76 rocksdb, 10 llvm-project).
3. `qualifiedNameText` treated any `::` in the text as a qualifier, so `Vec<std::string>` would have been stamped `prov="final-segment"` once (1) was fixed.

Reproduced as reported. On #268's head, `--callees=fillU` (`VecImpl<Decl *> &v`) declines, while `fillQ` (`ll::VecImpl<Decl *> &v`) splits to both `TBase::push_back`.

## Change

- `lastNameNode`: walk the grammar's own fields (a qualified name's `name`, then a template-id's `name`) to the last name, for written types and constructor names alike.
- `writtenTypeOf` accepts `template_type`. `scoped_type_identifier` leaves the accepted kinds, because no grammar that reaches this path (cpp/cuda/objc `parser.c`) has that symbol.
- `qualifiedNameText` asks the tree whether a scope was written, so a `::` inside a template argument qualifies nothing.
- The stale "class-name filter in buildGraph" comment above `ctorNameNode` is corrected: no such filter exists, and a non-class record tombstones.

**Built, measured, dropped: an unqualified template-function constructor (`auto v = Vec<T>()`).** That spelling is every cast helper. Accepting it recorded `dyn_cast`/`cast` as the type of `auto *CI = dyn_cast<CallInst>( I )` and `Spec = cast<FunctionDecl>( F )`, tombstoning their written types. On llvm-project that was 994 more retargeted sites, **779 of them lost edges**; on rocksdb 49, 35 lost. It is now a stated floor (arm 39d). The qualified `llvm::cast<T>( x )` still records `cast`, as on main; a follow-up task covers that.

## Measured

`--pin-census --no-cache`, C rows joined on (caller id, callee, line).

| population | target-changed sites | bound calls |
| --- | --- | --- |
| rocksdb `0e2801ac3`, main → this | 640 | 200,036 → 200,383 (+347) |
| llvm-project `4d5358b1d`, main → this | 5,231 | 1,126,051 → 1,129,640 (+3,589) |
| rocksdb, #268 `e5ca33b5` → #268 + this (scratch merge, never pushed) | 657 | 201,085 → 201,432 (+347) |
| llvm-project, #268 → #268 + this | 18,050 | 1,153,998 → 1,166,700 (+12,702) |

**Graded.** A seeded uniform sample of 100 retargeted sites (seed 20260917; 20 / 30 / 12 / 38 across the four rows above). Seven independent readers graded it against source. Each saw the two answers as **A and B in random order**, and the key was withheld.

| population | better | same | worse | transitions (old → new) |
| --- | --- | --- | --- | --- |
| rocksdb main → this | 20 | 0 | 0 | 14 NONE→RIGHT, 4 WRONG→RIGHT, 1 PARTIAL→RIGHT, 1 WRONG→PARTIAL |
| llvm main → this | 30 | 0 | 0 | 20 NONE→RIGHT, 9 WRONG→RIGHT, 1 PARTIAL→RIGHT |
| rocksdb #268 → #268+this | 11 | 1 | 0 | 5 NONE→RIGHT, 3 WRONG→RIGHT, 3 PARTIAL→RIGHT, 1 PARTIAL→PARTIAL |
| llvm #268 → #268+this | 38 | 0 | 0 | 23 NONE→RIGHT, 8 WRONG→RIGHT, 7 NONE→PARTIAL |
| **all** | **99** | **1** | **0** | Wilson 95% for "better": [94.6, 99.8] |

The 7 NONE→PARTIAL are #268's template-family split (`SmallVectorImpl<uint32_t>&`: primary and `<T, true>` push_back, where a trivially copyable element selects only the latter). The one `same` is a pure-virtual `InternalIteratorBase<IndexValue>*` dispatch where both answers carry one foreign iterator.

**Every lost edge was read.** rocksdb has none; llvm-project has nine (the #268 composition has the same nine):
- **7:** a name declared twice in one function with different types (`APInt Mask` beside `SmallVector<int> Mask`, `SmallString<128> Dir` beside a loop's `directory_iterator Dir`). Rule 2's flat per-function table cannot tell which declaration covers a call, so it drops both, as it already does for two plain types.
- **2:** `auto Table = llvm::EytzingerTable<…>::create( … )` and `auto SCCI = scc_iterator<…>::begin( … )`. These recorded the class only because the cut stopped at `<`; they now read `create`/`begin`, as the non-template `Foo::create()` always did.

Wall time, one `--no-cache --pin-census` run each on llvm-project under shared load: 7.35 s main, 7.40 s this (noise-level, not a benchmark).

## Gates

Red first, `test/narrowcheck.sh` arms 39–43 (fixture generated in-gate):
- **main 13a19162:** 9 FAIL — 39a–c and 40a (no edge), 40b (`prov=[NO-EDGE]`), 41a–c (`size@vecs/adt.h:28`, the wrong class), 43 (`mech=[NO-CENSUS-ROW]`).
- **#268 head e5ca33b5:** those 9 plus 42 ×3 — `fillU/fillL/fillN(): push_back -> [no edge], its qualified twin fillQ() -> [push_back@vecs/adt.h:13 push_back@vecs/adt.h:19]`.
- **A variant that reads qualification off the whole spelling:** 40b alone, `prov=[final-segment], want [none]`.
- **A variant that accepts the template-function constructor:** 39d is the only arm whose answer changes.
- **Green:** this branch and the #268 composition, including #268's arms 26–38.

Targeted, against this branch's binary, each ALL PASS: narrowcheck, cacheidentitycheck, chacheck, chaconecheck, cppqualcheck, declinecheck, decltodefcheck, fieldnarrowcheck, fieldusescheck, fnptrcheck, gateexitcheck, localitycheck, manifestcheck, pincensuscheck, recallevalcheck, resolverhonestycheck, shadowcheck, showcasecapturecheck, usescheck, qschemetripcheck, qextractionkeycheck.

Also checked:
- determinism: rocksdb map and census byte-identical across two runs;
- `xmllint --noout` clean;
- warm census == cold, and a cache written by the main binary is not served stale;
- `--quality-delta=$(git merge-base origin/main HEAD)..HEAD`: `gating="0" regressions="0"`.

Full suite, one run, foreground-equivalent (nothing else running), on the merged tree 50b97b7b: `gates=639 pass=636 skip=3 fail=0 wall=832.3s jobs=6 tree_writes=0` — the three environmental skips (argvdiffcheck, editchecknotecheck, g1freshcheck with no `asan/` in the worktree; no local ASan per the lane brief). `docs/gatecount_build.py --check` (625) and `docs/limits_build.py --check` clean.

## Pins moved

- `src/ingest_cache.h` `kParserVer` 99 → 103, with `src/quality.h` `kIngestParserVerMirror` (the train assigns the number).
- `test/qschemetrip.hash` re-pinned via `UPDATE_GOLDEN=1 test/qschemetripcheck.sh`, with a RE-PIN LOG entry.
- No legend, help, gate-count or printf-parity pin moved.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved C++ template-type handling for receiver bindings, including unqualified templates such as `Vec<T>` and nested qualified types such as `Outer<int>::Inner`.
  - Prevented `::` inside template arguments from causing incorrect qualification.
  - Restored correct bindings for template-typed parameters, locals, constructors, and range-based variables.
  - Avoided incorrect bindings to outer classes or same-named local symbols.

- **Tests**
  - Added coverage for template receiver resolution and related edge cases.

### CODE REVIEW

File: test/qschemetripcheck.sh [hunk 1]
```diff
@@ -34,6 +34,12 @@ SRC="$ROOT/src/quality.h"
 ING="$ROOT/src/ingest_cache.h"   # extraction-identity constants moved here (2026-08-29 ingest.cpp section split); the hashed CONCAT label keeps its historical spelling so the pin holds
 PIN="$ROOT/test/qschemetrip.hash"
 # RE-PIN LOG (the pin is a bare hash, so its justification has to live here).
+# 2026-09-17, TEMPLATE ARGUMENTS IN A RECEIVER'S TYPE (test/narrowcheck.sh arms 39-43): parserVer and its quality mirror move
+#   99 -> 103 — a C++ declaration's Type/ParamType record takes its type's last name through the grammar's fields, so an
+#   unqualified template-id records a type at last and `Outer<int>::Inner` records `Inner`, not `Outer`. 103 is declared past
+#   the 100-102 the lanes queued ahead of this one declare; the landing train assigns the number and re-derives this pin on the
+#   merged tree. Record layouts are unchanged, so kCacheVersion stays 22; kQSnapCacheScheme stays 12: no key or snapshot
```
[the remark this answers is not in the document]
Comment (joyful-ii-V-I) [reply test/qschemetripcheck.sh]: Valid — fixed in 97b499f0. The entry copied "stays 12" from the older RE-PIN LOG rows, which record the value current at the time; `src/quality.h:2888` declares `kQSnapCacheScheme = 14` on this tree (train 1: #253 → 13, #255 → 14). It is a comment only: the hashed manifest is quality.h / ingest_cache.h text, not this gate, so `test/qschemetrip.hash` does not move (`qschemetripcheck`, `gateexitcheck`, `manifestcheck`, `qextractionkeycheck` ALL PASS).

**Comment by joyful-ii-V-I** [comment]
This PR is part of integration train 3, #281, and will show as merged when that train lands.
